### PR TITLE
Support Yup nullable derived types

### PR DESCRIPTION
**Pull request by mauricedb**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: See below
- [x] ~Increase the version number in the header if appropriate.~
- [x] ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

It would be really nice if we can derive a TypeScript type alias from the Yup definition instead of having to redefine it. Using the `ReturnType` of the `cast()` function this is possible. The problem with this is that nullable defintions are lost. With this PR these are preserved.

Given the following Yup schema:
``` TypeScript
const personSchema = yup.object({
    firstName: yup.string(), // $ExpectType StringSchema<string>
    email: yup
        .string()
        .nullable()
        .email(),
    birthDate: yup
        .date()
        .nullable()
        .min(new Date(1900, 0, 1)),
    canBeNull: yup.string().nullable(true), // $ExpectType StringSchema<string | null>
    mustBeAString: yup
        .string()
        .nullable(true)
        .nullable(false),
    children: yup.array(yup.string()).nullable()
});
```

We can derive the `Person` type as follows:
``` TypeScript
type Person = ReturnType<typeof personSchema.cast>;
```

This is with this PR equivalent to:
``` TypeScript
type Person = {
    firstName: string;
    email: string | null;
    birthDate: Date | null;
    canBeNull: string | null;
    mustBeAString: string;
    children: string[] | null;
}
```

Without this PR it would lose the null types and be equivalent to:
``` TypeScript
type Person = {
    firstName: string;
    email: string;
    birthDate: Date;
    canBeNull: string;
    mustBeAString: string;
    children: string[];
}
```

Now we can create a new `Person` object as follows:
``` TypeScript
const person: Person = {
    firstName: "",
    email: null,
    birthDate: null,
    canBeNull: null,
    mustBeAString: "",
    children: null
};
```

### CODE REVIEW

File: types/yup/yup-tests.ts
Comment (sseppola) [inline types/yup/yup-tests.ts]: Would be nice to also ensure the properties can still be non-null, for example with some assignments like:
```
person.email = 'some@email.com';
person.birthDate = new Date();
person.children = ['1', '2', '3'];
```

**Review by sseppola — COMMENTED** [review]
This looks like 🔥, love that we can derive the type from the schema with nullables!

As far as I can tell this PR is 👌, I'd just like a second opinion from another maintainer (@YashdalfTheGray perhaps?), and for the tests to confirm non-null values.

File: types/yup/index.d.ts
Comment (YashdalfTheGray) [inline types/yup/index.d.ts]: I think, here, instead of defaulting it to string, we should say that `T extends string` because we're always going to want the user of `StringSchema` to pass us something that is derived from string specifically.

File: types/yup/index.d.ts
Comment (YashdalfTheGray) [inline types/yup/index.d.ts]: Same thing here, I think we should change `T = number` to `T extends number` because the second way locks the client from passing in some other type into this that might not be a number altogether.

File: types/yup/index.d.ts
Comment (YashdalfTheGray) [inline types/yup/index.d.ts]: Same here, `T extends Date`. This one matters less because users are _unlikely_ to have their own class derived from `Date` but it happens sometimes.

File: types/yup/index.d.ts
Comment (YashdalfTheGray) [inline types/yup/index.d.ts]: I'm guessing some of these changes are like prettier changes?

File: types/yup/yup-tests.ts
Comment (YashdalfTheGray) [inline types/yup/yup-tests.ts]: Maybe have a boolean test case as well since we're adding types for `BooleanSchema` too?

File: types/yup/yup-tests.ts
Comment (YashdalfTheGray) [inline types/yup/yup-tests.ts]: It would also be cool to add type assertions to this as well so that we're checking one of each for the nullable case!

**Review by YashdalfTheGray — CHANGES_REQUESTED** [review]

**Comment by YashdalfTheGray** [comment]
> This looks like 🔥, love that we can derive the type from the schema with nullables!
> 
> As far as I can tell this PR is 👌, I'd just like a second opinion from another maintainer (@YashdalfTheGray perhaps?), and for the tests to confirm non-null values.

This is definitely clutch, have a couple of suggestions to add more type safety. Most 🔥part is `Schema<Exclude<T, null>>`! TIL!

File: types/yup/index.d.ts
Comment (mauricedb) [reply types/yup/index.d.ts]: Yes sorry, I have prettier with format on save enabled. You want me to revert these?

File: types/yup/index.d.ts
Comment (mauricedb) [reply types/yup/index.d.ts]: @YashdalfTheGray Looks like I need to change the definition to `interface StringSchema<T extends string | null = string> extends Schema<T>` because it can be null as well as a string type. And defaults to string is none specified.

File: types/yup/yup-tests.ts
Comment (mauricedb) [reply types/yup/yup-tests.ts]: 👍

File: types/yup/index.d.ts
Comment (mauricedb) [reply types/yup/index.d.ts]: 👍

File: types/yup/yup-tests.ts
Comment (mauricedb) [reply types/yup/yup-tests.ts]: 👍

**Comment by mauricedb** [comment]
@YashdalfTheGray I am also looking at doing the same around `notRequired()`

File: types/yup/index.d.ts
Comment (YashdalfTheGray) [reply types/yup/index.d.ts]: I'm okay with this. What do y'all think? I like that we are forcing the client of this library to pass in something that has known string methods so that the validator will work properly but it might be too restrictive possibly? Thoughts?

File: types/yup/index.d.ts
Comment (YashdalfTheGray) [reply types/yup/index.d.ts]: Oh no. I like prettier being run on the code!

**Comment by YashdalfTheGray** [comment]
> @YashdalfTheGray I am also looking at doing the same around `notRequired()`

[Aww yiss](https://media.giphy.com/media/VtcjgulX9Z6dW/giphy.gif)

File: types/yup/index.d.ts
Comment (sseppola) [reply types/yup/index.d.ts]: I think its fair to enforce string type here.
I didn't think of this before, but does 
`interface StringSchema<T extends string | null = string> extends Schema<T>`
prevent me from passing undefined into the schema? if so, should it be 
`interface StringSchema<T extends string | null | undefined = string> extends Schema<T>` instead?

Comment (mauricedb) [reply types/yup/index.d.ts]: @sseppola Yes it should prevent that. When using `--strictNullChecks` both `null` and `undefined` are distinctly different types. But that said when would you pass undefined into the `StringSchema` interface? As far as I can tell that should never happen at the moment. It will when I do the second PR with `notRequired()`/`required()` but I will do that as a separate PR.

Comment (sseppola) [reply types/yup/index.d.ts]: I tried to make a codesandbox to explore the default behavior of Yup:
https://codesandbox.io/s/formik-yup-form-dskdh

In summary
````
import * as Yup from "yup";

async function validate(value) {
  try {
    const res = await Yup.string().isValid(value);
    console.log(`"${value}" is valid`);
  } catch (error) {
    console.log(`"${value}" is not valid`);
  }
}
validate(undefined);
validate(null);
validate("");

```
the output is:
```
"undefined" is valid
"null" is not valid
"" is valid
```

So to be in line with Yup the type should be even more complicated:
`interface StringSchema<T extends string | null | undefined = string | undefined> extends Schema<T>`

However, as we go from `interface StringSchema extends Schema<string>`, which enforces string, then what you have currently is fine, and we can deal with `undefined` in another PR :)

**Review by sseppola — APPROVED** [review]

**Review by YashdalfTheGray — APPROVED** [review]

**Comment by sandersn** [comment]
I guess that it would a more efficient to manually clone the Schema types, something like `StringSchema` and `StringNullableSchema`. Type parameters are a very general mechanism to turn a type into a factory of types, but you only ever need to construct two variants. 

I think it would also be easier to read for users of the library (especially JS users) to see `s.nullable() : StringNullableSchema` instead of `s.nullable() : StringSchema<string | null>`.

**Comment by mauricedb** [comment]
@sandersn I am sure that copying the schema's to have a `StringSchema` and `StringNullableSchema` would be more efficient right now. However I would prefer not to do so. For one it would make removing `null` quite a bit harder with the `.nullable(false)` operation. And secondly it would result in just two copies of the schema now but in the next PR I am adding support for `undefined` through the `.required()` and `.notRequired()` operations exploding the possible schema's to: `StringSchema`, `StringNullableSchema`, `StringUndefinedSchema` and `StringNullableAndUndefinedSchema`. And that is just for `string`, the same happens for other types. Just imagine what happens if another option turns up, not that I am expecting one 😄

**Comment by sandersn** [comment]
Yeah, four subtypes for every original type is a bit much. It might be easier to read if Schema took an extra `Nullable extends null | undefined = never` parameter and `StringSchema<Nullable = null | undefined extends never> extends Schema<string, Nullable>`, so that people would see `StringSchema<null>` or `StringSchema<never>`. But I'm not sure that's less confusing, actually.

**Comment by YashdalfTheGray** [comment]
What if we used parameterized interfaces? Like `Nullable<string>`, that way, it could let us compose them like `Nullable<Undefinable<string>>` or something like that. I do like the idea of making these aliases though so we're not putting a huge thing in the parameter argument.

**Comment by amcasey** [comment]
@YashdalfTheGray @mauricedb Do you want to revise this or merge it as-is?

**Comment by YashdalfTheGray** [comment]
I'll defer to @mauricedb for it. I know he's got another PR in the works so I'm confident that these changes will make it into the repo in the near future. I think if he doesn't reply, a fallback is to just merge this as it is.

**Comment by mauricedb** [comment]
Sorry for the delay, been a busy week with two conferences.

I would prefer to merge this as is. I am doing a second PR after this to add required/notRequired with undefined and will see if about updating the generic types and how that affects `Schema<Exclude<T, null>>`.

**Comment by YashdalfTheGray** [comment]
🎉 Nice! @amcasey, we'll just merge this as is then!

**Pull request merged**
